### PR TITLE
bsc#1120717: Check all aliases for etcd member check and create /etc/caasp directory before access

### PR DESCRIPTION
**Pull request by bergmannf**
This PR adds fixes that were needed to make an update from a v2 to a v3 cluster succeed:

- generating the `/etc/hosts` file failed, because the backup location did not exist.

- checking `etcd` membership would (wrongly) fail, because only the list of aliases for either the `ipv4` or `ipv6` addresses would be checked, instead of both.

### CODE REVIEW

**Comment by drpaneas** [comment]
Doesn't seem to work

```
Feb 12 17:34:46 admin salt-minion[2006]: [ERROR   ] Error while bringing up minion for multi-master. Is master at localhost responding?
Feb 12 17:35:20 admin salt-minion[2006]: [ERROR   ] Failed to import module caasp_net, this is due most likely to a syntax error:
Feb 12 17:35:20 admin salt-minion[2006]: Traceback (most recent call last):
Feb 12 17:35:20 admin salt-minion[2006]:   File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1459, in _load_module
Feb 12 17:35:20 admin salt-minion[2006]:     mod = imp.load_module(mod_namespace, fn_, fpath, desc)
Feb 12 17:35:20 admin salt-minion[2006]:   File "/var/cache/salt/minion/extmods/modules/caasp_net.py", line 47
Feb 12 17:35:20 admin salt-minion[2006]:     return found_aliases
Feb 12 17:35:20 admin salt-minion[2006]:          ^
Feb 12 17:35:20 admin salt-minion[2006]: SyntaxError: invalid syntax
Feb 12 17:35:23 admin salt-minion[2006]: [ERROR   ] State 'caasp_hosts.managed' was not found in SLS 'etc-hosts'
Feb 12 17:35:23 admin salt-minion[2006]: Reason: 'caasp_hosts.managed' is not available.
Feb 12 17:35:45 admin salt-minion[2006]: [ERROR   ] Failed to import module caasp_net, this is due most likely to a syntax error:
Feb 12 17:35:45 admin salt-minion[2006]: Traceback (most recent call last):
Feb 12 17:35:45 admin salt-minion[2006]:   File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1459, in _load_module
Feb 12 17:35:45 admin salt-minion[2006]:     mod = imp.load_module(mod_namespace, fn_, fpath, desc)
Feb 12 17:35:45 admin salt-minion[2006]:   File "/var/cache/salt/minion/extmods/modules/caasp_net.py", line 47
Feb 12 17:35:45 admin salt-minion[2006]:     return found_aliases
Feb 12 17:35:45 admin salt-minion[2006]:          ^
Feb 12 17:35:45 admin salt-minion[2006]: SyntaxError: invalid syntax
Feb 12 17:35:47 admin salt-minion[2006]: [ERROR   ] State 'caasp_hosts.managed' was not found in SLS 'etc-hosts'
Feb 12 17:35:47 admin salt-minion[2006]: Reason: 'caasp_hosts.managed' is not available.
Feb 12 17:37:10 admin salt-minion[2006]: [ERROR   ] Rendering exception occurred
Feb 12 17:37:10 admin salt-minion[2006]: Traceback (most recent call last):
Feb 12 17:37:10 admin salt-minion[2006]:   File "/usr/lib/python2.7/site-packages/salt/utils/templates.py", line 170, in render_tmpl
Feb 12 17:37:10 admin salt-minion[2006]:     output = render_str(tmplstr, context, tmplpath)
Feb 12 17:37:10 admin salt-minion[2006]:   File "/usr/lib/python2.7/site-packages/salt/utils/templates.py", line 391, in render_jinja_tmpl
Feb 12 17:37:10 admin salt-minion[2006]:     buf=tmplstr)
Feb 12 17:37:10 admin salt-minion[2006]: SaltRenderError: Jinja variable 'salt.utils.templates.AliasedLoader object' has no attribute 'caasp_nodes'
Feb 12 17:37:10 admin salt-minion[2006]: [ERROR   ] Unable to render top file: Jinja variable 'salt.utils.templates.AliasedLoader object' has no attribute 'caasp_nodes'
Feb 12 17:40:33 admin salt-minion[2006]: [WARNING ] /usr/lib/python2.7/site-packages/salt/modules/mine.py:296: DeprecationWarning: the target type should be passed using the 'tgt_type' argument instead of 'expr_form'. Support for using 'expr_form' will be removed in Salt Fluorine.
Feb 12 17:40:35 admin salt-minion[2006]: [ERROR   ] /etc/hosts was not generated successfully: _makedirs() got an unexpected keyword argument 'exist_ok'
```

File: salt/_modules/caasp_hosts.py [hunk 1]
```diff
@@ -68,6 +68,19 @@ def __virtual__():
...
 
 
+def _makedirs(path):
+    try:
+        # Python3 allows to pass exist_ok=True to not throw an error
+        # Python2 does not have this parameters, so we have to catch the OSError
+        os.makedirs(path)
+    except OSError as e:
+        # Error 17 = file exists, so nothing for us to do
+        if e.errno == 17:
```
Comment (MalloZup) [inline salt/_modules/caasp_hosts.py]: We could log a warning  like file exists etc

File: salt/_modules/caasp_net.py [hunk 2]
```diff
@@ -36,11 +36,12 @@ def get_aliases(hostname):
...
     """
     hosts = get_hosts()
+    found_aliases = set([hostname])
     for host, aliases in hosts.iteritems():
         if hostname in aliases:
-            __utils__['caasp_log.debug']('CaaS: retrieved aliases %s for %s', aliases, hostname)
-            return aliases
-    return [hostname]
+            found_aliases.update(set(aliases))
+    debug('CaaS: retrieved aliases %s for %s', aliases, hostname)
```
Comment (MalloZup) [inline salt/_modules/caasp_net.py]: Fix typo caas should be caasp instead of caas (even if not from your pr)

File: salt/_modules/caasp_hosts.py [hunk 3]
```diff
@@ -68,6 +68,19 @@ def __virtual__():
     return "caasp_hosts"
 
 
+def _makedirs(path):
```
Comment (MalloZup) [inline salt/_modules/caasp_hosts.py]: We should call actually this function make dir singular because it Is creating only 1dir from a var not list

**Comment by mjura** [comment]
@bergmannf It has to be rebased

File: salt/_modules/caasp_net.py (hunk 2 above)
Comment (MalloZup) [reply salt/_modules/caasp_net.py]: agreed thx , https://github.com/kubic-project/salt/issues/740

**Comment by MalloZup** [comment]
@bergmannf the master branch here should be caasp V4, which AFAIK doesn't support migration from to 2 to 3 and other combinations, so IMHO should be only a 30 branch pr. Or I am missing something? Tia

**Comment by bergmannf** [comment]
@MalloZup I detected this during the updates from v2 -> v3, but the changes in the codes should also make the code more robust in the v4 release (especially using all the aliases is something that should go into v4 as well).

**Comment by MalloZup** [comment]
@bergmannf backport to 30 if needed thx. before merging it, put the bug number in the commit msg

**Comment by MalloZup** [comment]
the PR/CI is failing on `update`. it is a known race condition on our salt states. retriggering

**Comment by bergmannf** [comment]
@MalloZup CI finally passed :+1:

**Comment by MalloZup** [comment]
@bergmannf  to me is good to merge. Unless somebody have some things against it. @inercia cc

**Pull request merged**